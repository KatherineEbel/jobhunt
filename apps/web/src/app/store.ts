import {combineReducers, configureStore, PreloadedState} from '@reduxjs/toolkit'
import { jobHuntApi } from 'services/jobHuntApi'
import authReducer from 'features/auth/authSlice'
import jobsReducer from 'features/jobs/jobsSlice'
import alertReducer from 'features/alert/alertSlice'

const rootReducer = combineReducers({
  [jobHuntApi.reducerPath]: jobHuntApi.reducer,
  alerts: alertReducer,
  auth: authReducer,
  jobs: jobsReducer,
})

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(jobHuntApi.middleware),
    preloadedState,
    reducer: rootReducer,
  })
}

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']