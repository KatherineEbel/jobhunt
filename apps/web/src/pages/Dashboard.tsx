import { useAppContext } from 'context/appContext'
import { useEffect } from 'react'
import { useNavigate } from 'react-router-dom'

const Dashboard = () => {
  const { user } = useAppContext()
  const navigate = useNavigate()
  useEffect(() => {
    if (!user) {
      navigate('/register', { replace: true })
    }
  }, [user])

  if (!user) return null

  return (
    <div>
      <h1>Dashboard</h1>
      <p>Welcome {user.name}</p>
    </div>
  )
}

export default Dashboard
