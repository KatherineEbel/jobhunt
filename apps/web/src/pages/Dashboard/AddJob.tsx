import {useAppContext} from 'context/appContext'
import {CreateJobRequest} from 'lib'
import {useMemo} from 'react'
import {useNavigate, useSearchParams} from 'react-router-dom'
import { JobForm} from 'ui'

const AddJob = () => {
  const { addJob, editJob, jobs } = useAppContext()
  const navigate = useNavigate()

  const [searchParams] = useSearchParams()
  const jobId = searchParams.get('jobId')

  const job = useMemo(() => {
    if (!jobId) return undefined
    const match = jobs.find(j => j.id === jobId)
    if (match) {
      const {position, location, company, status, contract} = match
      return {position, location, company, status, contract}
    }
  }, [jobId])

  const handleSubmit = async (values: CreateJobRequest) => {
    console.log(values)
    const success = jobId && job ? await editJob(jobId, values) : await addJob(values)
    if (success) {
      navigate('/jobs')
    }
    return success
  }
  return <>
    <h1>{jobId ? 'Edit' : 'Add'} Job</h1>
    <JobForm onSubmit={handleSubmit} job={job} />
  </>
}

export default AddJob
