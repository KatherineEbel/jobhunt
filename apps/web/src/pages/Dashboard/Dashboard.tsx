import { useToggle } from 'hooks/useToggle'
import { Nav } from 'Nav'
import React from 'react'
import { Outlet } from 'react-router-dom'
import { SharedLayout, SmallSidebar, LargeSidebar } from 'ui'

const Dashboard = () => {
  const { open, toggleOpen } = useToggle()

  return (
    <SharedLayout>
      <main className="dashboard">
        <SmallSidebar onClose={toggleOpen} open={open} />
        <LargeSidebar onClose={toggleOpen} open={open}/>
        <div>
          <Nav toggleSidebar={toggleOpen} />
          <div className="dashboard-page">
            <Outlet />
          </div>
        </div>
      </main>
    </SharedLayout>
  )
}

export default Dashboard
