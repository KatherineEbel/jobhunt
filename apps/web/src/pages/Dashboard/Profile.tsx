import {useAppContext} from 'context/appContext'
import { ProfileForm, ProfileValues, ProfilePageWrapper } from 'ui'

const Profile = () => {
  const { updateUser, user } = useAppContext()
  if (!user) return null

  const [firstName, lastName] = user.name.split(' ')
  const onSubmit = async (values: ProfileValues) => {
    await updateUser(values)
  }
  return (
    <ProfilePageWrapper>
      <h1>Profile Page</h1>
      <ProfileForm onSubmit={onSubmit} user={{firstName, lastName}}/>
    </ProfilePageWrapper>
  )
}

export default Profile
