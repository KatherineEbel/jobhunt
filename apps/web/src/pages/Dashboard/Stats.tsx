// import { useAppContext } from 'context/appContext'
// import { useEffect } from 'react'
// import { useNavigate } from 'react-router-dom'

const Stats = () => {
  // const { user } = useAppContext()
  // const navigate = useNavigate()
  // useEffect(() => {
  //   if (!user) {
  //     navigate('/register', { replace: true })
  //   }
  // }, [user])
  //
  // if (!user) return null
  //
  return (
    <div>
      <h1>Stats</h1>
      <p>Welcome</p>
    </div>
  )
}

export default Stats
