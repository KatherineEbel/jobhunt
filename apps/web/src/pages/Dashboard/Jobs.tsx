import {createAlert} from 'features/alert/alertSlice'
import {useAppDispatch} from 'hooks/store'
import {JobQuery} from 'lib'
import {useState} from 'react'
import {useDeleteJobMutation, useJobsQuery} from 'services/jobHuntApi'
import {JobList, Loader, SearchForm, SearchFormValues} from 'ui'

const Jobs = () => {
  const dispatch = useAppDispatch()
  const [query, setQuery] = useState<Partial<JobQuery>>({})
  const { data, error, isLoading, refetch } = useJobsQuery(query)
  const [deleteJob, {isError}]  = useDeleteJobMutation()

  if (error) {
    dispatch(createAlert({type: 'danger', message: 'Failed to fetch jobs'}))
    console.log(error)
  }

  if (isError) {
    dispatch(createAlert({type: 'danger', message: 'Failed to delete job'}))
  }

  const onSubmit = (values: SearchFormValues) => {
    const { status, contract} = values
    setQuery({
      ...values,
      status: status === 'all' ? undefined : status,
      contract: contract === 'all' ? undefined : contract
    })
    refetch()
  }

  const clearFilters = () => {
    setQuery({})
    refetch()
  }
  
  if (isLoading) return <Loader/>

  return <>
    <SearchForm onSubmit={onSubmit} onReset={clearFilters}/>
    <JobList jobs={data?.jobs || []}  onDeleteJob={deleteJob}/>
  </>
}

export default Jobs
