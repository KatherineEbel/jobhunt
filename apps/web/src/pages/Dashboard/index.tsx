import AddJob from 'pages/Dashboard/AddJob'
import Jobs from 'pages/Dashboard/Jobs'
import Profile from 'pages/Dashboard/Profile'
import Stats from 'pages/Dashboard/Stats'
import Dashboard from 'pages/Dashboard/Dashboard'

export { AddJob, Dashboard, Jobs, Profile, Stats }
