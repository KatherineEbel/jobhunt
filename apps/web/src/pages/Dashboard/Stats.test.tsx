import {waitForElementToBeRemoved} from '@testing-library/react'
import {setupStore} from 'app/store'
import {setCredentials} from 'features/auth/authSlice'
import {Stats} from 'pages/Dashboard/index'
import {PropsWithChildren} from 'react'
import {renderWithProviders, screen} from 'testUtils'

const user = {
  firstName: 'Johnny',
  lastName: 'Smith',
  location: 'New York City',
  email: 'johnny@example.com',
  token: '62e9b9de742aed9193d4e23f',
  createdAt: new Date().toUTCString(),
  updatedAt: new Date().toUTCString(),
}

jest.mock('recharts', () => {
  const OriginalModule = jest.requireActual('recharts')
  return {
    ...OriginalModule,
    ResponsiveContainer: ({children}: PropsWithChildren) => (
      <OriginalModule.ResponsiveContainer width={800} height={800}>
        {children}
      </OriginalModule.ResponsiveContainer>
    )
  }
})

describe('Stats Page', function () {
  beforeEach(() => {
    window.ResizeObserver = jest.fn().mockImplementation(() => ({
      observe: jest.fn(),
      unobserve: jest.fn(),
      disconnect: jest.fn(),
    }));
  })
  test('render', async () => {
    const store = setupStore()
    store.dispatch(setCredentials({user}))
    const {asFragment} = renderWithProviders(<Stats/>, { store })
    await waitForElementToBeRemoved(() => screen.getByTestId('loader'))
    expect(screen.getByText(/20/)).toBeInTheDocument()
    expect(screen.getByText(/15/)).toBeInTheDocument()
    expect(screen.getByText(/22/)).toBeInTheDocument()
    expect(screen.getByText(/20/)).toBeInTheDocument()
    expect(asFragment()).toMatchSnapshot()
  })
})