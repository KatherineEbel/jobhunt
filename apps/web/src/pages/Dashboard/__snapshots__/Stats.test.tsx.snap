// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stats Page render 1`] = `
<DocumentFragment>
  <div>
    <section
      class="sc-cCsOjp ksVKhq"
    >
      <article
        class="sc-gXmSlM knfiHX"
        color="#e9b949"
      >
        <header>
          <p
            class="count"
          >
            20
          </p>
          <div
            class="icon"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M336 160H48c-26.51 0-48 21.49-48 48v224c0 26.51 21.49 48 48 48h16v16c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16v-16h128v16c0 8.84 7.16 16 16 16h32c8.84 0 16-7.16 16-16v-16h16c26.51 0 48-21.49 48-48V208c0-26.51-21.49-48-48-48zm-16 216c0 4.42-3.58 8-8 8H72c-4.42 0-8-3.58-8-8v-16c0-4.42 3.58-8 8-8h240c4.42 0 8 3.58 8 8v16zm0-96c0 4.42-3.58 8-8 8H72c-4.42 0-8-3.58-8-8v-16c0-4.42 3.58-8 8-8h240c4.42 0 8 3.58 8 8v16zM144 48h96v80h48V48c0-26.51-21.49-48-48-48h-96c-26.51 0-48 21.49-48 48v80h48V48z"
              />
            </svg>
          </div>
        </header>
        <h3
          class="title"
        >
          pending applications
        </h3>
      </article>
      <article
        class="sc-gXmSlM jIFzZZ"
        color="#647acb"
      >
        <header>
          <p
            class="count"
          >
            18
          </p>
          <div
            class="icon"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M436 160H12c-6.627 0-12-5.373-12-12v-36c0-26.51 21.49-48 48-48h48V12c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v52h128V12c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v52h48c26.51 0 48 21.49 48 48v36c0 6.627-5.373 12-12 12zM12 192h424c6.627 0 12 5.373 12 12v260c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V204c0-6.627 5.373-12 12-12zm333.296 95.947l-28.169-28.398c-4.667-4.705-12.265-4.736-16.97-.068L194.12 364.665l-45.98-46.352c-4.667-4.705-12.266-4.736-16.971-.068l-28.397 28.17c-4.705 4.667-4.736 12.265-.068 16.97l82.601 83.269c4.667 4.705 12.265 4.736 16.97.068l142.953-141.805c4.705-4.667 4.736-12.265.068-16.97z"
              />
            </svg>
          </div>
        </header>
        <h3
          class="title"
        >
          interviews scheduled
        </h3>
      </article>
      <article
        class="sc-gXmSlM KnuZZ"
        color="#d66a6a"
      >
        <header>
          <p
            class="count"
          >
            22
          </p>
          <div
            class="icon"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M511.988 288.9c-.478 17.43-15.217 31.1-32.653 31.1H424v16c0 21.864-4.882 42.584-13.6 61.145l60.228 60.228c12.496 12.497 12.496 32.758 0 45.255-12.498 12.497-32.759 12.496-45.256 0l-54.736-54.736C345.886 467.965 314.351 480 280 480V236c0-6.627-5.373-12-12-12h-24c-6.627 0-12 5.373-12 12v244c-34.351 0-65.886-12.035-90.636-32.108l-54.736 54.736c-12.498 12.497-32.759 12.496-45.256 0-12.496-12.497-12.496-32.758 0-45.255l60.228-60.228C92.882 378.584 88 357.864 88 336v-16H32.666C15.23 320 .491 306.33.013 288.9-.484 270.816 14.028 256 32 256h56v-58.745l-46.628-46.628c-12.496-12.497-12.496-32.758 0-45.255 12.498-12.497 32.758-12.497 45.256 0L141.255 160h229.489l54.627-54.627c12.498-12.497 32.758-12.497 45.256 0 12.496 12.497 12.496 32.758 0 45.255L424 197.255V256h56c17.972 0 32.484 14.816 31.988 32.9zM257 0c-61.856 0-112 50.144-112 112h224C369 50.144 318.856 0 257 0z"
              />
            </svg>
          </div>
        </header>
        <h3
          class="title"
        >
          jobs declined
        </h3>
      </article>
      <article
        class="sc-gXmSlM jsaOrd"
        color="#2fbb2f"
      >
        <header>
          <p
            class="count"
          >
            15
          </p>
          <div
            class="icon"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 640 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M255.03 261.65c6.25 6.25 16.38 6.25 22.63 0l11.31-11.31c6.25-6.25 6.25-16.38 0-22.63L253.25 192l35.71-35.72c6.25-6.25 6.25-16.38 0-22.63l-11.31-11.31c-6.25-6.25-16.38-6.25-22.63 0l-58.34 58.34c-6.25 6.25-6.25 16.38 0 22.63l58.35 58.34zm96.01-11.3l11.31 11.31c6.25 6.25 16.38 6.25 22.63 0l58.34-58.34c6.25-6.25 6.25-16.38 0-22.63l-58.34-58.34c-6.25-6.25-16.38-6.25-22.63 0l-11.31 11.31c-6.25 6.25-6.25 16.38 0 22.63L386.75 192l-35.71 35.72c-6.25 6.25-6.25 16.38 0 22.63zM624 416H381.54c-.74 19.81-14.71 32-32.74 32H288c-18.69 0-33.02-17.47-32.77-32H16c-8.8 0-16 7.2-16 16v16c0 35.2 28.8 64 64 64h512c35.2 0 64-28.8 64-64v-16c0-8.8-7.2-16-16-16zM576 48c0-26.4-21.6-48-48-48H112C85.6 0 64 21.6 64 48v336h512V48zm-64 272H128V64h384v256z"
              />
            </svg>
          </div>
        </header>
        <h3
          class="title"
        >
          offers received
        </h3>
      </article>
    </section>
    <section
      class="sc-ciZhAO iCtmaE"
    >
      <h1>
        Monthly Applications
      </h1>
      <button
        class="sc-iBkjds dACIzx"
      >
        View Area Chart
      </button>
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 300px;"
      />
    </section>
  </div>
</DocumentFragment>
`;
